### PR TITLE
Add cheats in per game configuration 

### PR DESCRIPTION
**Pull request by luc-git**
Hello with my pull request you can add cheat with per game property but the problem is that it only works if a game is running.
If it's doesn't the case citra will crash with an access violation exception nullptr. So this pointer will point to the running game if I understand.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6379)

### CODE REVIEW

**Comment by SleepingSnakezzz** [comment]
Sorry, I don't understand the point of this. If the cheat codes can only be added whilst the game is launched (just like the normal cheats menu), then all you've done here is add extra steps to accessing the same cheats menu that users can already access by `Emulation>Cheats`.

**Comment by luc-git** [comment]
Yes of course but I think it would be a good idea to add cheats when a game is not running

**Comment by luc-git** [comment]
It could be possible because other emulators can do this way

**Review by FearlessTobi — CHANGES_REQUESTED** [review]
I see no benefit from this, if it introduces no new functionality.
As Snake said, Cheats already can be configured when a game is running.

**Comment by luc-git** [comment]
Yes this is the problem of the cheats configuration it only works if a game is running

**Comment by SleepingSnakezzz** [comment]
Then the appropriate way to suggest this is to open a feature request in our Github's issues section. Opening a PR with a non-functional change does not help.

**Comment by GPUCode** [comment]
I think this PR has merit, cheats should ideally go to the per-game menu. But the crashing issue needs to be fixed first. Consider disabling the tab if the menu is opened without a game running.

**Review by GPUCode — CHANGES_REQUESTED** [review]
While the Qt code looks fine, duplicating the cheat management functions in the GUI side isn't the correct solution here. This results in having two different implementations that do the same thing.

**Comment by luc-git** [comment]
So can I remove functions from the core side?

**Comment by luc-git** [comment]
If I use the core part there is an access violation

**Comment by luc-git** [comment]
You need to execute this line https://github.com/citra-emu/citra/blob/master/src/core/core.cpp#L312 to prevent the accessviolation.
But if I want to pass this  https://github.com/citra-emu/citra/blob/master/src/citra_qt/main.cpp#L2593 it will not work

**Comment by luc-git** [comment]
Maybe we need to pass the pointer

**Comment by GPUCode** [comment]
My suggestion; add a constructor to System and move the creation of CheatEngine there. Then update LoadCheatFile and SaveCheatFile to receive the program_id instead of taking it from system directly. Downside is that this requires quite a bit of work updating CheatDialog and android to account for that but I believe it's the best solution.
Maybe wait until another developer can suggest something else, someone could have a simpler idea

**Comment by luc-git** [comment]
Can program_id have a value even if a game is not running?

**Comment by GPUCode** [comment]
It does not, but the per game dialog has access to the program id of the game being configured, which is why the cheat engine needs to be modified to receive it

**Comment by luc-git** [comment]
> While the Qt code looks fine, duplicating the cheat management functions in the GUI side isn't the correct solution here. This results in having two different implementations that do the same thing.

Fixed

File: src/citra_qt/cheats.cpp [hunk 1]
```diff
@@ -217,10 +206,15 @@ void CheatDialog::OnDeleteCheat() {
     }
 
     edited = false;
-    ui->buttonSave->setEnabled(false);
     ui->buttonAddCheat->setEnabled(true);
 }
 
+bool CheatDialog::ApplyConfiguration() {
+    if (edited)
```
Comment (GPUCode) [inline src/citra_qt/cheats.cpp:213]: Add braces

File: src/citra_qt/configuration/configure_per_game.cpp [hunk 2]
```diff
@@ -81,6 +87,11 @@ void ConfigurePerGame::ResetDefaults() {
     close();
 }
 
+void ConfigurePerGame::SaveCheat() {
+    if (cheat_tab->ApplyConfiguration())
```
Comment (GPUCode) [inline src/citra_qt/configuration/configure_per_game.cpp:91]: Ditto for braces

File: src/core/cheats/cheats.cpp [hunk 3]
```diff
@@ -19,9 +19,11 @@ namespace Cheats {
 // we use the same value
 constexpr u64 run_interval_ticks = 50'000'000;
 
-CheatEngine::CheatEngine(Core::System& system_) : system(system_) {
+CheatEngine::CheatEngine(u64 title_id_, Core::System& system_)
+    : system(system_), title_id{title_id_} {
     LoadCheatFile();
-    Connect();
+    if (Core::System::GetInstance().KernelRunning())
```
Comment (GPUCode) [inline src/core/cheats/cheats.cpp:25]:
```suggestion
    if (system.IsPoweredOn())
```

**Review by GPUCode — COMMENTED** [review]
Looks good overall now, great job!

**Comment by luc-git** [comment]
Changes are resolved any news?

**Comment by SleepingSnakezzz** [comment]
I only have a couple small issues:
image

Red: "Cheat" should be renamed to "Cheats". 

Blue: The `Apply` button is **not** there when opening per-game settings from Citra's game list (before launching your game). It is only available *after* the game is launched through `Emulation>Configure Current Game...`
It should be available for both. 

Also, you should probably rename the `Apply` button to `Save`. Since applying a cheat code is not its function, saving it is.

Lastly, I don't like the `Apply` button's current position, nor the fact that it is visible in other settings tabs, since it has no use there. I would suggest to move it to position [1] or [2] as I've shown in the picture, so that it is specific to the `Cheats` tab (I prefer position [2]).

**Comment by SleepingSnakezzz** [comment]
Other than the things I mentioned above, I couldn't find any other issues on the desktop side of things. Awesome work!

**Comment by luc-git** [comment]
Another thing is that cheats are applied even without saving it in the cheat file

**Comment by luc-git** [comment]
As soon as a cheat is added in https://github.com/citra-emu/citra/blob/2a2ee8bc968232058fe77a06e95d4f3fa1b851e7/src/core/cheats/cheats.h#L39 it is applied in game

**Comment by luc-git** [comment]
Maybe it could be better to apply cheats only after clicking save or ok button

File: src/citra_qt/configuration/configure_cheats.h [hunk 4]
```diff
@@ -5,22 +5,23 @@
 #pragma once
 
 #include <memory>
-#include <QDialog>
+#include "core/cheats/cheats.h"
```
Comment (FearlessTobi) [inline src/citra_qt/configuration/configure_cheats.h:8]: this seems like it can be a forward declaration

**Comment by luc-git** [comment]
> Lastly, I don't like the Apply button's current position, nor the fact that it is visible in other settings tabs, since it has no use there. I would suggest to move it to position [1] or [2] as I've shown in the picture, so that it is specific to the Cheats tab (I prefer position [2]).

I don't understand why move Apply button position to delete button position?

**Comment by SleepingSnakezzz** [comment]
> The apply button is not cheat specific, it has been there since per game settings were implemented. I believe it's meant as a way to adjust per game settings while the game is running, without closing the window. So imo that fix is to per-game settings in general and out of scope of this PR.
Correct. I was mixing things up (though I still don't see why this button needs to exist, but that's a different matter entirely).

@luc-git The reason why I was going on about that button is because when a user wants to enter multiple cheat codes at the same time, the `Apply` button is the only way to do so without having to constantly reopen the settings. And this button is **not** available when you open the per-game settings from the game list (before you launch your game).

For example, if I want to add 3 cheats *before* I launch a game, the only way to do so is to click `Add Cheat`, fill in the cheat code and then press `Ok`. After that, I have to reopen the per game settings again, and do that 2 more times. This is because the `Apply` button does not exist there.

It makes more sense to me to have a `Save Cheat` button next to the `Delete` or `Add Cheat` button (the positions that I marked in my screenshot) for this purpose.

File: src/citra_qt/configuration/configure_cheats.h (hunk 4 above)
Comment (luc-git) [reply src/citra_qt/configuration/configure_cheats.h]: Unfortunately if I remove the cheats.h include the github workflow fails with macos and windows build. But it works for me I use visual studio IDE in windows it's weird.

**Comment by luc-git** [comment]
@SleepingSnakezzz I think is good I added save button again in the cheat tab

**Comment by luc-git** [comment]
Closes: #6380

**Comment by SleepingSnakezzz** [comment]
Tried the PR artifact, looks good to me!

**Comment by luc-git** [comment]
Ok but I don't know why forward declare CheatEngine doesn't work

**Comment by luc-git** [comment]
@SleepingSnakezzz Workflow compiles successfully for windows but not for macos if I replace include by a forward declaration. I had to disable ccache to compile for windows

**Comment by luc-git** [comment]
> ```
> error: unknown type name 'u64'
>     explicit ConfigureCheats(u64 title_id_, QWidget* parent = nullptr);
>                              ^
> /Users/runner/work/citra/citra/src/citra_qt/configuration/configure_cheats.h:58:5: error: unknown type name 'u64'
>     u64 title_id;
>     ^
> ```
> 
> You need to include `common/common_types.h` Edit: you might also need to `#include<vector>`

Ok thank you very much it build now. Yes it's weird it compile even if I don't include vector. But there is a std::vector declaration in the header file.

File: src/core/core.cpp [hunk 5]
```diff
@@ -311,7 +311,8 @@ System::ResultStatus System::Load(Frontend::EmuWindow& emu_window, const std::st
         }
     }
     kernel->SetCurrentProcess(process);
-    cheat_engine = std::make_unique<Cheats::CheatEngine>(*this);
+    cheat_engine = std::make_unique<Cheats::CheatEngine>(
+        Kernel().GetCurrentProcess()->codeset->program_id, *this);
```
Comment (FearlessTobi) [inline src/core/core.cpp:315]: the title id is being read already in line 317. you can just use the information from there

File: src/citra_qt/configuration/configure_per_game.cpp [hunk 6]
```diff
@@ -81,13 +87,24 @@ void ConfigurePerGame::ResetDefaults() {
...
     general_tab->ApplyConfiguration();
     system_tab->ApplyConfiguration();
     enhancements_tab->ApplyConfiguration();
     graphics_tab->ApplyConfiguration();
     audio_tab->ApplyConfiguration();
     debug_tab->ApplyConfiguration();
+    if (ui->tabWidget->currentIndex() == ui->tabWidget->indexOf(cheat_tab.get()) &&
+        buttonSender == ui->buttonBox->button(QDialogButtonBox::Apply)) {
+        cheat_tab->ApplyConfiguration();
+    };
```
Comment (FearlessTobi) [inline src/citra_qt/configuration/configure_per_game.cpp:107]:
```suggestion
    }
```

File: src/citra_qt/configuration/configure_per_game.cpp [hunk 7]
```diff
@@ -81,13 +87,24 @@ void ConfigurePerGame::ResetDefaults() {
...
 }
 
+void ConfigurePerGame::SaveCheat() {
+    if (cheat_tab->ApplyConfiguration()) {
+        accept();
+    }
+}
+
 void ConfigurePerGame::ApplyConfiguration() {
+    QPushButton* buttonSender = qobject_cast<QPushButton*>(sender());
```
Comment (FearlessTobi) [inline src/citra_qt/configuration/configure_per_game.cpp:97]: I really don't like this way of doing this.
also the variable name is not compliant with our coding style

**Review by FearlessTobi — CHANGES_REQUESTED** [review]

File: src/citra_qt/configuration/configure_per_game.cpp (hunk 7 above)
Comment (luc-git) [reply src/citra_qt/configuration/configure_per_game.cpp:97]: Ok but why is this variable name using the same coding style? https://github.com/citra-emu/citra/blob/f96047f18229e257297593e68cbbdf29b72b1e9b/src/citra_qt/main.cpp#L1709

**Comment by luc-git** [comment]
Done but now if you press ok button cheats will only be applied if you are in the cheat tab.
This is to prevent qmessage from cheats to be opened while we are in another tab.

**Comment by luc-git** [comment]
Hello changes requested are resolved anything new?

**Review by FearlessTobi — APPROVED** [review]

**Comment by SleepingSnakezzz** [comment]
As this PR has said, it's now been moved to the per-game settings. Regardless, this isn't something to ask on the PR itself. Questions related to Citra should be asked on the Citra forums or in the Citra Discord server.

**Pull request merged**